designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,439 @@
+
+#### ● A new `compareResultsByFilePath` static method
+
+This method receives two lint results then returns `+1`, `-1`, or `0`. This method is intended to use in order to sort results.
+
+<details>
+<summary>A rough sketch of the `compareResultsByFilePath()` method.</summary>
+
+```js
+class ESLint {
+  static compareResultsByFilePath(a, b) {
+    if (a.filePath < b.filePath) {
+      return -1
+    }
+    if (a.filePath > b.filePath) {
+      return 1
+    }
+    return 0
+  }
+}
+```
+
+</details>
+
+### ■ Deprecate `CLIEngine` class
+
+This RFC soft-deprecates `CLIEngine` class.
+
+Because it's tough to maintain two versions (sync and async) of implementation. The two are almost copy-pasted stuff, but hard to share the code. Therefore, this RFC deprecates the sync version to improve our code with the way which needs asynchronous behavior in the future. For example, `CLIEngine` cannot support [linting in parallel](https://github.com/eslint/rfcs/pull/42), plugins/configs as ES modules, etc...
+
+### ■ Out of scope
+
+- Not change API for rules. This RFC doesn't change APIs that rule implementation uses. We may be able to support asynchronous stuff in rules in the future, but it's out of this RFC's scope.
+- Not change internal logics. This RFC just adds the public interface that is asynchronous. It would be a wrapper of `CLIEngine` for now.
+
+## Documentation
+
+- The "[Node.js API](https://eslint.org/docs/developer-guide/nodejs-api)" page should describe the new public API and deprecation of `CLIEngine` class.
+
+## Drawbacks
+
+People that use `CLIEngine` have to update their application with the new API. It will need hard work.
+
+## Backwards Compatibility Analysis
+
+This is a breaking change.
+
+Deprecating `CLIEngine` is a drastic change. But people can continue to use `CLIEngine` as-is until we decide to remove it.
+
+The new API depends on [Asynchronous Iteration](https://github.com/tc39/proposal-async-iteration) syntax. Node.js supports the syntax since `10.0.0`, so we have to drop Node.js `8.x`. Because the `8.x` will be EOL in December 2019 (two months later!), we can work on this soon.
+
+## Alternatives
+
+- Adding `engine.executeAsyncOnFiles()`-like methods and we maintain it along with the existing synchronous API. But as what I wrote in the "[Deprecate `CLIEngine` class](#deprecate-cliengine-class)" section, it would be tough.
+- Using [Streams](https://nodejs.org/dist/latest-v12.x/docs/api/stream.html) instead of [Asynchronous Iteration](https://github.com/tc39/proposal-async-iteration). We can introduce `ESLint` class in a minor release if we used Streams. But because Node.js 8 will be EOL two months later, we should be able to use Asynchronous Iteration soon. And Streams don't support `for-await-of` syntax until `v11.14.0`. Iterator protocol is smaller spec than streams, and it's easy to use.
```

Comment:
Asynchronous iteration ought to be a much friendlier approach than streams. :+1: Given the timing, requiring Node.js v10+ is not an issue as you mentioned above.